sdk/metric/exemplar/reservoir.go

```diff
@@ -11,6 +11,7 @@ import (
 )
 
 // Reservoir holds the sampled exemplar of measurements made.
+// All methods must be safe to invoke concurrently.
```

Comment:
Is this backwards compatible?

Reply:
What if they implement something like a `concurrentSafe` interface:

```golang
type concurrentSafe interface {
	ConcurrentSafe() bool
}
```

And we skip the wrapping?

Reply:
Also, how did you determine wrapping a user reservoir is 6x slower? Aren't we doing the same thing by adding centralized locking in the storeage?

Reply:
I actually like that it appears in package comments.  Makes it more discoverable for users that might be confused about performance.

Reply:
> Thats based on #7427, and making the assumption that we should be able to get very little lock contention for reservoirs that have capacity == num CPU and proper locking strategies. The ExemplarDisabled benchmark for counters is ~75ns, and the ExemplarEnabled benchmark is ~300ns because of locking.
> 
> Edit: I didn't account for the cost of exemplar recording itself, but it is still a significant cost.

I still don't follow this logic. If exemplars are enable and we are wrapping a user implementation with locking we will see results like you show in #7427, not 6x the results you show in #7427, right?

Reply:
> I actually like that it appears in package comments. Makes it more discoverable for users that might be confused about performance.

I'm not sure that is true. The `ConcurrentSafe` type is a distinct entry in the API and documentation. It is an added responsibility to link the two concepts for a user. Ultimately the "concurrent safe interface" that a reservoir can implement needs to be documented in the `Reservoir` documentation. If that is the case I don't see the added benefit of adding to the API.

Not exporting the type will also mean the feature can be prototyped without any API changes. Which will be ideal if we ultimately decide to not commit to this approach.